8267140: Support closing the HttpClient by making it auto-closable

dfuch opened the pull request:
Please find here an RFE that makes the `java.net.HttpClient` auto-closeable.

The API has been modeled on `ExecutorService`.

HttpClient::close() is a graceful shutdown and will wait until all operations are terminated before returning.
If a request is in progress, and the caller doesn't pull the corresponding data (for instance, the request was sent with a BodyHandler.ofInputStream(), but the caller stopped reading the input stream) then close() may never return.

Therefore, additional methods similar to those present in `ExecutorService` are also proposed. In summary:

- `shutdown()`: initiate a graceful shutdown, but doesn't wait for termination. 
- `shutdownNow()`:  initiate an immediate shutdown, attempting to cancel all operations in progress. Doesn't wait for termination.
- `awaitTermination(Duration)`: await for termination within the given delay
- `isTerminated()` tells whether the client is terminated

New tests have been added to test the proposed behaviors.

HttpClient tests (new and old) are still stable.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8304165](https://bugs.openjdk.org/browse/JDK-8304165) to be approved

### Issues
 * [JDK-8267140](https://bugs.openjdk.org/browse/JDK-8267140): Support closing the HttpClient by making it auto-closable
 * [JDK-8304165](https://bugs.openjdk.org/browse/JDK-8304165): Support closing the HttpClient by making it auto-closable (**CSR**)


### Reviewers
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13019/head:pull/13019` \
`$ git checkout pull/13019`

Update a local copy of the PR: \
`$ git checkout pull/13019` \
`$ git pull https://git.openjdk.org/jdk.git pull/13019/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13019`

View PR using the GUI difftool: \
`$ git pr show -t 13019`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13019.diff">https://git.openjdk.org/jdk/pull/13019.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13019#issuecomment-1468208653)

dfuch commented:
/csr needed

dfuch commented:
Here is the CSR draft: https://bugs.openjdk.org/browse/JDK-8304165

turbanoff commented on test/jdk/java/net/httpclient/ShutdownNow.java:
```diff
@@ -0,0 +1,385 @@
+        Throwable accepted = null;
+        while (t != null) {
+            out.println(what + ": checking " + t);
+            if (t instanceof  IOException io && hasExpectedMessage(io)) {
```
```suggestion
            if (t instanceof IOException io && hasExpectedMessage(io)) {
```

turbanoff commented on test/jdk/java/net/httpclient/AsyncShutdownNow.java:
```diff
@@ -0,0 +1,452 @@
+        Throwable accepted = null;
+        while (t != null) {
+            out.println(what + ": checking " + t);
+            if (t instanceof  IOException io && hasExpectedMessage(io)) {
```
```suggestion
            if (t instanceof IOException io && hasExpectedMessage(io)) {
```

turbanoff commented on test/jdk/java/net/httpclient/HttpClientShutdown.java:
```diff
@@ -0,0 +1,476 @@
+        Throwable accepted = null;
+        while (t != null) {
+            out.println(now() + what + ": checking " + t);
+            if (t instanceof  IOException io && hasExpectedMessage(io)) {
```
```suggestion
            if (t instanceof IOException io && hasExpectedMessage(io)) {
```

turbanoff commented on test/jdk/java/net/httpclient/HttpClientShutdown.java:
```diff
@@ -0,0 +1,476 @@
+                out.println(now() + "Start sending body for: " + t.getRequestURI());
+                for (int i=0; i<bytes.length; i++) {
+                    long now = System.nanoTime();
+                    long sincePrevious =  Duration.ofNanos(now - previous).toMillis();
```
```suggestion
                    long sincePrevious = Duration.ofNanos(now - previous).toMillis();
```

Michael-Mc-Mahon commented on src/java.net.http/share/classes/java/net/http/HttpClient.java:
```diff
@@ -730,4 +741,121 @@ public enum Redirect {
+     *
+     * @since 21
+     */
+    public void shutdownNow() {
```
The first sentence of the `shutdownNow` api doc should describe the behavior of the method itself, ie that it does an immediate shutdown.

Also, I wonder if there is another way to formulate these implSpec comments? Naive readers might think they have to override the method.

dfuch replied:
I have added some text to the `@apiNote` and added a link in all the `@implSpec`. Does that read better?

jaikiran replied:
I think this part in this method's javadoc is a bit out of place:

> It is called if the thread waiting on {@link #close()} is interrupted.

It is just one of the place from where this method will be called from. Since the `shutdownNow()` is a public API, it could be called by applications too. Perhaps we should just remove that sentence from that method's javadoc? The javadoc of `close()` explains in a better way that this `shutdownNow()` method might get invoked from within `close()`, which I think should be enough.

dfuch replied:
Right. In my first prototype I had made this method protected, until I realized it was actually better to make it public. This is a remnant of that first prototype. I will remove this sentence here, since as you noticed it's already specified at in close().

dfuch commented:
Thanks Andrey @turbanoff for the suggestions. I have applied them.

jaikiran commented on src/java.net.http/share/classes/java/net/http/HttpClient.java:
```diff
@@ -130,9 +131,23 @@
+ * An {@code HttpClient} instance typically manages its own pools of
+ * connections, which it can typically reuse when another request is made to
+ * the same origin server using the same protocol version and security,
+ * through the same client. In the JDK implementation, connection pools
```
Hello Daniel, do you think we should maybe skip these details on how the connection pool decides to use a connection from pool? Perhaps, just make a mention that the HttpClient instance manages connections in a pool and may reuse it as and when necessary?

dfuch replied:
OK

jaikiran commented on src/java.net.http/share/classes/java/net/http/HttpClient.java:
```diff
@@ -130,9 +131,23 @@
+ * the same origin server using the same protocol version and security,
+ * through the same client. In the JDK implementation, connection pools
+ * are not shared between {@code HttpClient} instances. Creating a new client
+ * for each operation, though possible, will prevent reusing such connections.
```
I think, the reason why this sentence about creating a new connection for each operation, is here is to discourage applications from doing that. However, I found it a bit odd to have it in the middle of a section named "closing". 
The existing javadoc of this `HttpClient` has this sentence:
> Once built, an {@code HttpClient} is immutable, and can be used to send multiple requests.

Perhaps we should add this sentence there and emphasise that creating a new client for each operation is discouraged since that prevents reuse of existing connections?

dfuch replied:
Good suggestion. I'm moving that in the paragraph following the one you quoted, since that paragraph already talks about sharing resources. I have softened the wording though, since the text is moved from a non-normative section to a normative section.

jaikiran commented on src/java.net.http/share/classes/java/net/http/HttpClient.java:
```diff
@@ -730,4 +745,126 @@ public enum Redirect {
+
+    /**
+     * Initiates an orderly shutdown in which previously submitted
+     * operations are completed, but no new request will be accepted.
```
I think the word "completed" can be confusing, given its usage in `java.util.concurrent.Future` APIs. I haven't yet reviewed the code, but as far as I understand, the shutdown implementation merely marks a state which won't allow any new requests to be created, but let's the current submitted operations run to completion. The implementation however doesn't (mark such operations) as completed. Do you think we should word this differently? Maybe "Initiates an orderly shutdown in which previously submitted operations are allowed to run to completion, but no new request will be accepted"?

dfuch replied:
Good idea. Done.

jaikiran commented on src/java.net.http/share/classes/java/net/http/HttpClient.java:
```diff
@@ -730,4 +745,126 @@ public enum Redirect {
+
+    /**
+     * Blocks until all operations have completed execution after a shutdown
+     * request, or the timeout occurs, or the current thread is
```
If I understand this correctly, this method is expected to be called after `shutdown()` or `shutdownNow()` has been called. If so, then should this method throw an `IllegalStateException` if it gets called before any of those shutdown methods are called?

jaikiran replied:
Would reword this sentence to:

> Blocks until all operations have completed execution after a shutdown request, or the {@code duration} passes, or the current thread is {@linkplain Thread#interrupt() interrupted}, whichever happens first.

dfuch replied:
> should this method throw an `IllegalStateException` if it gets called before any of those shutdown methods are called?

ExecutorService doesn't do that, and I don't think we should here either. The actual implementation of the method just joins the selector manager thread. There is no requirement that shutdown() and awaitTermination() are performed in the same thread, and no technical reason to prevent waiting before shutdown (especially if shutdown is triggered by some other thread).

> Do you think we could reword this sentence to, to have necessary pointers to duration and thread interruption:

Good idea.

jaikiran commented on src/java.net.http/share/classes/java/net/http/HttpClient.java:
```diff
@@ -730,4 +745,126 @@ public enum Redirect {
+     * Subclasses should override this method to implement the proper behavior.
+     * See the API Note on {@linkplain ##closing closing}.
+     *
+     * @return {@code true} if all tasks have completed following shut down
```
Same here. I think it should be "following a shutdown"

dfuch commented on src/java.net.http/share/classes/java/net/http/HttpClient.java:
```diff
@@ -730,4 +745,126 @@ public enum Redirect {
+     * Subclasses should override this method to implement the proper behavior.
+     * See the API Note on {@linkplain ##closing closing}.
+     *
+     * @param duration the maximum time to wait
```
Good point. I borrowed this paragraph from `Thread::join`:

```
     * <p> This method does not wait if the duration to wait is less than or
     * equal to zero. In this case, the method just tests if the thread has
     * terminated.
```

dfuch commented on src/java.net.http/share/classes/java/net/http/HttpClient.java:
```diff
@@ -730,4 +745,126 @@ public enum Redirect {
+     * interrupted, whichever happens first.
+     *
+     * @implSpec
+     * The default implementation of this method does nothing and returns true.
```
If that where the case, then `close()` would never return when none of these methods are overridden (case of existing subclasses).  `isTerminated` stipulates that it will never return `true` if shutdown has not been called, which is borrowed from `ExecutorService` specification, and maintaining a state in this super class is not really a good idea. The choice was being throwing an exception (that would make close throw too), or return `true`, which would allow close() to trivially exit if not implemented. All in all I believe I prefer to have `awaitTermination` returning `true` by default.

jaikiran commented on src/java.net.http/share/classes/java/net/http/HttpClient.java:
```diff
@@ -646,7 +660,8 @@ public enum Redirect {
      * <ul>
-     * <li>{@link IOException} - if an I/O error occurs when sending or receiving</li>
+     * <li>{@link IOException} - if an I/O error occurs when sending or receiving,
+     *      or the client {@linkplain ##closing shutting down}.</li>
```
Hello Daniel, there's a "is" missing here. Should have been "client is ..."

jaikiran commented on test/jdk/java/net/httpclient/AsyncShutdownNow.java:
```diff
@@ -0,0 +1,452 @@
+    static Throwable cleanup(HttpClient client, ExecutorService readerService, Throwable failed) {
+        try {
+            try {
+                if (client.awaitTermination(Duration.ofMillis(2000))) {
```
Hello Daniel, given our past experience of such timeouts leading to intermittent failures (on slow systems), do you think we should perhaps use a timeout duration of one day here? That way we don't have to guess the right timeout value and the jtreg test framework can timeout this test if it executes past the configured test timeout.

dfuch replied:
If it fails I'd like to do some analysis of the logs before just increasing the timeout.

jaikiran replied:
That sounds reasonable.

jaikiran commented on test/jdk/java/net/httpclient/AsyncShutdownNow.java:
```diff
@@ -0,0 +1,452 @@
+                try (OutputStream os = t.getResponseBody()) {
+                    String msg = "Incorrect uuid header values:[" + uuids + "]";
+                    (new RuntimeException(msg)).printStackTrace();
+                    t.sendResponseHeaders(500, -1);
```
Given that we are writing out a response body (on the next line), should the second param here be the size of those bytes, instead of `-1`?

dfuch replied:
Here -1 only means that the body will be transmitted in chunk mode for HTTP/1, and that an additional empty DataFrame with length=0 and the FIN bit set will be sent after the message bytes when the output stream is closed for HTTP/2 - so I don't think it really matters. Both modes should work.

jaikiran commented on test/jdk/java/net/httpclient/HttpClientBuilderTest.java:
```diff
@@ -39,6 +39,7 @@
 import java.time.Duration;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
```
The copyright year on this `HttpClientBuilderTest.java` file needs an update.

jaikiran commented on test/jdk/java/net/httpclient/HttpClientClose.java:
```diff
@@ -0,0 +1,345 @@
+            failed = cleanup(readerService, failed);
+        }
+        if (failed instanceof Exception ex) throw ex;
+        if (failed instanceof Error e) throw e;
```
Is this missing a `client.isTerminated()` assertion here?

dfuch replied:
Well spotted.

jaikiran commented on test/jdk/java/net/httpclient/HttpClientClose.java:
```diff
@@ -0,0 +1,345 @@
+    static Throwable cleanup(ExecutorService readerService, Throwable failed) {
+        try {
+            readerService.shutdown();
+            readerService.awaitTermination(2000, TimeUnit.MILLISECONDS);
```
Same comment about the timeout value.

dfuch replied:
Same answer: there should be no operation in progress at this point - since we have waited for all the completable futures to complete, and we're not relying on the GC to garbage things.

jaikiran commented on test/jdk/java/net/httpclient/HttpClientClose.java:
```diff
@@ -0,0 +1,345 @@
+                try (OutputStream os = t.getResponseBody()) {
+                    String msg = "Incorrect uuid header values:[" + uuids + "]";
+                    (new RuntimeException(msg)).printStackTrace();
+                    t.sendResponseHeaders(500, -1);
```
Same comment about the `-1` value.

dfuch replied:
Right - same answer: there's no harm in using chunked mode.

jaikiran commented on test/jdk/java/net/httpclient/offline/FixedResponseHttpClient.java:
```diff
@@ -240,4 +259,54 @@ public void onSubscribe(Flow.Subscription subscription) {
+        Duration max = Duration.ofMillis(Long.MAX_VALUE);
+        long timeout = duration.compareTo(max) > 0 ? Long.MAX_VALUE : duration.toMillis();
+        try {
+            all.exceptionally((t) -> null).get(timeout, TimeUnit.MILLISECONDS);
```
This implementation of `awaitTermination` only waits for the current (snapshot) of requests that are currently in progress. If a new request happens to be initiated (and is in-progress) after this snapshot of `responses.toArray()` was taken, then this implementation of `awaitTermination` can possibly return an incorrect `true` (indicating the client has terminated) while there currently might be some requests in progress. This `FixedResponseHttpClient` is just a test specific implementation which appears to be used in one single test class (`OfflineTesting`) and as far as I can see, that test won't be impacted with the implementation of `awaitTermination()`, so I think this implementation might be OK for now.

dfuch replied:
Good point.

dfuch replied:
OK - this ended up to be a bit more complicated than I thought. What I had missed is that the `FixedResponseHttpClient` actually wraps another (real) `HttpClient`, so close etc... need to close both. I have now fixed this and updated the offline test to close its clients too.

jaikiran commented:
The latest updated wording to the specification in `HttpClient` looks good to me. It adds more clarity on what kind of work might happen (internally in the background) when the client is closed. The implementation changes too look good to me. I've added some test specific comments inline.

Overall, the changes in this PR look good to me and I believe the addition of these new APIs will allow applications to control the lifecycle and resources used by the `HttpClient` in a much more deterministic way.

jaikiran commented on test/jdk/java/net/httpclient/offline/DelegatingHttpClient.java:
```diff
@@ -112,4 +112,34 @@ public <T> HttpResponse<T> send(HttpRequest request,
+
+    /**
+     * Invokes the default implementation of HttpClient::close
+     * Can be closed by subclasses if necessary.
```
I think this is a typo in the javadoc and I suspect you meant "Can be overridden by subclasses if necessary."

jaikiran commented on test/jdk/java/net/httpclient/offline/FixedResponseHttpClient.java:
```diff
@@ -40,19 +45,27 @@
  * Suitable for testing where network connections are to be avoided.
+ * Can delegate to an actual HttpClient created from a supplied
+ * HttpClient.Builder if needed, by calling methods on its
+ * DelegatingClient super class.
```
Minor typo, should have been `DelegatingHttpClient` instead of `DelegatingClient`

jaikiran reviewed (approved):
There a couple of trivial typos in comments, which I have noted inline. The `OfflineTesting` file would need a copyright year update.

Other than that, the latest changes in `6226acf` all look good to me.

dfuch commented:
@AlanBateman suggested the following changes, with which I have updated the CSR and this PR:

In the class level documentation:

1. Only the first sentence of the previously added `@apiNote` remains in the `@apiNote`
2. The bulk of the `@apiNote` text that was added previously is now moved to the `@implNote` paragraph
3. The modified `@apiNote` is moved above the `@implNote` that was previously there
4. The `##closing` anchor now links to the `@implNote`

In each of the new methods:

1. The link to the `##closing` paragraph (non-normative) is removed from the `@implSpec` (normative).
2. An `@see ##closing Implementation Note on closing the HttpClient` is added instead, at the end of the method API documentation.

jaikiran commented:
Hello Daniel, the latest documentation changes all look fine to me.

As for:

>
>     2. An `@see ##closing Implementation Note on closing the HttpClient` is added instead, at the end of the method API documentation.

I haven't previously used this construct in a `@see`, so as long as the generated javadoc renders fine, this looks good. Does it render a `<a href=...>Implementation Note on closing the HttpClient</a>`?

dfuch commented:
/integrate

dfuch commented:
>I haven't previously used this construct in a @see, so as long as the generated javadoc renders fine, this looks good. Does it render a <a href=...>Implementation Note on closing the HttpClient</a>?

yes - I have generated the apoi doc and it looks fine
